### PR TITLE
client: Implement custom map control with sign-in/sign-out

### PR DESCRIPTION
**Pull request by mikol**
- Replace default Leaflet control
  - Add sign-in/sign-out to map control

screencast 2022-08-23 14-36-10

### CODE REVIEW

File: src/v2/components/FireListMapControl.jsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    <div className="c7e-map--control">
+      <ButtonGroup>
+        <IconButton icon='c7e-icon--zoom-in' label="Zoom in" title="Zoom in" onClick={() => map.zoomIn()}/>
+        <IconButton icon='c7e-icon--zoom-out' label="Zoom out" title="Zoom out" onClick={() => map.zoomOut()}/>
+      </ButtonGroup>
+
+      { isAuthenticated
+        ? <IconButton icon='c7e-icon--sign-out' label="Sign out" title="Sign out" onClick={signOut}/>
+        : <IconButton icon='c7e-icon--sign-in' label="Sign in" title="Sign in" onClick={signIn}/>
+      }
```
Comment (kinshuk) [inline src/v2/components/FireListMapControl.jsx:54]: I'm used to seeing login/logout at the top right.  Is your plan to move it later?

File: src/v2/App.jsx [hunk 2]
```diff
@@ -20,39 +20,64 @@ import AppFooter from './components/AppFooter.jsx'
...
+        }
+      })
+  }, [])
+
   const handleToggleAuthn = useCallback((title = null, fn = null, hide = false) => {
-    // XXX: Functions aren’t state. They return state. So wrap and return `fn`.
+    // XXX: Functions aren’t state. They return state. Wrap and return callback.
     // See: https://reactjs.org/docs/hooks-reference.html#functional-updates
-    setHandleAuthenticated(() => fn)
+    setHandleAuthenticated(() => () => {
```
[the remark this answers is not in the document]
Comment (mikol) [reply src/v2/App.jsx:47]: Looks like you may have deleted this comment? I don’t see it in the diffs… Anyway, `useState()` accepts a function argument, which it executes and uses the returned value as `state`. In this case, we would like `state` to be a function so the function argument returns a function.

File: src/v2/components/FireListMapControl.jsx (hunk 1 above)
Comment (mikol) [reply src/v2/components/FireListMapControl.jsx:54]: Yeah, I’m not sure. One reason to have it top-left is to be closest to the vote buttons (which I am considering swapping positions with the pagination buttons). That way the sign-out is immediately next to the most likely thing to trigger a sign-in. But, you’re absolutely correct, people will look for sign-in/-out in the top-right. So maybe we should put it there along with the settings/preferences button, and put the zoom controls bottom-right (where Google Maps has them—in both cases no less).

Google Maps 2022-08-25 17-59-02

File: src/v2/App.jsx [hunk 3]
```diff
@@ -20,39 +20,64 @@ import AppFooter from './components/AppFooter.jsx'
...
       /complete/.test(document.readyState)
         // XXX: Set scroll position asynchronously; otherwise, as observed,
         // scroll position setting isn’t guaranteed to take effect.
         ? requestAnimationFrame(() => window.scrollTo(0, 0))
         : setTimeout(check)
     })()
-  }, [])
+  }, [updateAuthentication])
+
+  console.log({isAuthenticated})
```
[the remark this answers is not in the document]
Comment (mikol) [reply src/v2/App.jsx:73]: Sorry, this is only for temporary debugging—I’m going to add an eslint setting to catch `console.*`. If something is going to be used for production logging, I generally make that explicit like so:

```js
const {error: report, log: log} = console
```

Or something like that.

**Review by kinshuk — APPROVED** [review]
After this PR, should voting code check isAuthenticated or continue to rely on API failure to trigger authentication?

**Comment by mikol** [comment]
I'd keep the failure case there no matter what.

ガラスカラスからのメッセージ。

**Pull request merged**